backend/src/models/reports/Tab_21_rpt_PA_Billing-Historical.js

```diff
@@ -1,16 +1,111 @@
+      "historical_projects.total_project_budget",
+      "fiscal_year.fiscal_year",
+      "historical_project_billing.q1",
+      "historical_project_billing.q2",
+      "historical_project_billing.q3",
+      "historical_project_billing.q4"
+    ),
+
+  projectRecovery: knex("fiscal_year")
+    .select({
+      project_id: "project.id",
+      project_number: "project.project_number",
+      project_name: "project.project_name",
+      total_project_budget: "project.total_project_budget",
+      fiscal_year: "fiscal_year.fiscal_year",
+      q1: knex.raw(`sum(case when quarter = '1' then amount else null end)`),
+      q2: knex.raw(`sum(case when quarter = '2' then amount else null end)`),
+      q3: knex.raw(`sum(case when quarter = '3' then amount else null end)`),
+      q4: knex.raw(`sum(case when quarter = '4' then amount else null end)`),
+      total_recovered: knex.sum("jv.amount"),
+    })
+    .innerJoin("jv", "fiscal_year.id", "jv.fiscal_year_id")
+    .innerJoin("project", "jv.project_id", "project.id")
+    .groupBy(
+      "project.id",
+      "project.project_number",
+      "project.total_project_budget",
+      "fiscal_year.fiscal_year"
+    ),
+};
+
+/**
+ * Retrieves final historical recoveries for the report.
+ *
+ * @returns {Knex.QueryBuilder} Knex query builder for fetching report data.
+ */
+const reportQueries = {
+  report: () =>
+    knex
+      .select({
+        budget_fiscal: "project_recovery_historical.budget_fiscal",
+        project_number: "project_recovery_historical.project_number",
+        project_name: "project_recovery_historical.project_name",
+        total_project_budget: "project_recovery_historical.total_project_budget",
+        q1: "project_recovery_historical.q1",
+        q2: "project_recovery_historical.q2",
+        q3: "project_recovery_historical.q3",
+        q4: "project_recovery_historical.q4",
+        total_recovered: knex.raw(
+          "cast( COALESCE(Q1, '0') + COALESCE(Q2, '0') + COALESCE(Q3, '0') + COALESCE(Q4, '0') as money)"
+        ),
+      })
+      .from(baseQueries.projectRecoveryHistorical.as("project_recovery_historical"))
+      .unionAll([
+        knex
+          .select({
+            budget_fiscal: "project_recovery.fiscal_year",
+            project_number: "project_recovery.project_number",
+            project_name: "project_recovery.project_name",
+            total_project_budget: "project_recovery.total_project_budget",
+            q1: "q1",
+            q2: "q2",
+            q3: "q3",
+            q4: "q4",
+            total_recovered: "total_recovered",
+          })
+          .from(baseQueries.projectRecovery.as("project_recovery"))
+          .orderBy("project_number"),
+      ]),
 };
 
 module.exports = {
   required: ["portfolio"],
-  getAll: async (query) => {
-    const { portfolio } = query;
-    const [report] = await Promise.all([getReport(portfolio)]);
+  getAll: async () => {
+    const [report] = await Promise.all([reportQueries.report()]);
```

Comment:
should this take a query and use that parameter (either fiscal year or portfolio)

to restrict the results?

something like:

```
module.exports = {
  required: ["fiscal"],
  getAll: async ({ fiscal }) => {
    const [{ fiscal_year }, report, report_totals] = await Promise.all([
      reportQueries.fiscalYear(fiscal),
      reportQueries.report(fiscal),
      reportQueries.report_totals(fiscal),
    ]);
    
    ```

Reply:
Thanks! There was some confusion due to inconsistent sample reports, I will restrict it to fiscal as you suggest, for now.